pkg/controller/machineset/machineset_controller.go

```diff
@@ -0,0 +1,208 @@
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Don't take action if the annotation is not present
+	annotations := instance.ObjectMeta.GetAnnotations()
+	if annotations == nil {
+		return reconcile.Result{}, nil
+	}
+	_, present := annotations[AutoScaleAnnotation]
+	if !present {
+		return reconcile.Result{}, nil
+	}
+
+	// Make sure the MachineSet has a non-empty selector.
+	msselector, err := metav1.LabelSelectorAsSelector(&instance.Spec.Selector)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if msselector.Empty() {
+		// The cluster-api machinesetcontroller expects every MachineSet to have
+		// its Selector set.
+		log.Info("MachineSet has empty selector, which is unexpected. Will not attempt scaling.")
+		return reconcile.Result{}, nil
+	}
+
+	hostselector, err := actuator.SelectorFromProviderSpec(&instance.Spec.Template.Spec.ProviderSpec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	hosts := &bmh.BareMetalHostList{}
+	opts := &client.ListOptions{
+		Namespace: instance.Namespace,
+	}
+
+	err = r.List(ctx, hosts, client.UseListOptions(opts))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var count int32
+	for i := range hosts.Items {
+		matches, err := r.hostMatches(ctx, hostselector, msselector, &hosts.Items[i])
+		switch {
+		case err == errConsumerNotFound:
+			return reconcile.Result{}, nil
+		case err != nil:
+			return reconcile.Result{}, err
+		case matches == true:
+			count++
+		}
+	}
+
+	if instance.Spec.Replicas == nil || count != *instance.Spec.Replicas {
+		log.Info("Scaling MachineSet", "new_replicas", count, "old_replicas", instance.Spec.Replicas)
+		new := instance.DeepCopy()
+		new.Spec.Replicas = &count
+		err = r.Update(ctx, new)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// hostMatches returns true if the BareMetalHost matches the MachineSet.
+func (r *ReconcileMachineSet) hostMatches(ctx context.Context, hostselector labels.Selector,
+	msselector labels.Selector, host *bmh.BareMetalHost) (bool, error) {
+	consumer := host.Spec.ConsumerRef
+
+	if consumer == nil {
+		return hostselector.Matches(labels.Set(host.ObjectMeta.Labels)), nil
+	}
+
+	// We will only count this host if it is consumed by a Machine that
+	// is part of the current MachineSet.
+	machine := &machinev1beta1.Machine{}
+	if consumer.Kind != "Machine" || consumer.APIVersion != machinev1beta1.SchemeGroupVersion.String() {
+		// this host is being consumed by something else; don't count it
+		return false, nil
+	}
+
+	// host is being consumed. Let's get the Machine and see if it
+	// matches the current MachineSet.
+	nn := types.NamespacedName{
+		Name:      consumer.Name,
+		Namespace: consumer.Namespace,
+	}
+	err := r.Get(ctx, nn, machine)
+	if err != nil {
+		if errors.IsNotFound(err) {
```

Comment:
Why is this considered such a catastrophic error instead of just a failure to match?

Reply:
This means there is a BareMetalHost with a ConsumerRef to a resource that doesn't exist. It could be that it was deleted after we read the BMH. Or it could be that the BMH is just in an unexpected state with a reference to a non-existent resource. Either way, it seems prudent to not make any scale changes until that gets back to a consistent state.